File: application/backend-credit-card-enrollment/backend-java/src/main/java/cloud/ambar/creditcard/enrollment/event/EnrollmentDeclined.java
package cloud.ambar.creditcard.enrollment.event;

import cloud.ambar.common.event.TransformationEvent;
import cloud.ambar.creditcard.enrollment.aggregate.Enrollment;
import cloud.ambar.creditcard.enrollment.aggregate.EnrollmentStatus;
import lombok.Getter;
import lombok.NonNull;
import lombok.experimental.SuperBuilder;

@SuperBuilder
@Getter
public class EnrollmentDeclined extends TransformationEvent<Enrollment> {
    @NonNull private String reasonCode;
    @NonNull private String reasonDescription;

    public Enrollment transformAggregate(Enrollment aggregate) {
        return aggregate.toBuilder()
                .aggregateVersion(aggregateVersion)
                .status(EnrollmentStatus.DECLINED.name())
                .build();
    }
}

File: application/backend-credit-card-enrollment/backend-c#/CreditCardEnrollment/Domain/Enrollment/Events/EnrollmentDeclined.cs
using CreditCardEnrollment.Common.Events;

namespace CreditCardEnrollment.Domain.Enrollment.Events;

public class EnrollmentDeclined : TransformationEvent
{
    public string UserId { get; set; } = string.Empty;
    public string ProductId { get; set; } = string.Empty;
    public string ReasonCode { get; set; } = string.Empty;
    public string ReasonDescription { get; set; } = string.Empty;
}
